File: Algorithmen/Primzahlen/Cpp/Primzahlen_1bisK.cpp
#include <iostream>
#include <cmath>
#include <ctime>

using namespace std;

int to_int(char* str) {
    int i = 0;
    while (*str != '\0') {
        i = i * 10 + (*str - '0');
        str++;
    }
    return i;
}

bool is_prime(int n) {
    if (n < 2) return false;
    for (int i = 2; i <= sqrt(n); i++) {
        if (n % i == 0) return false;
    }
    return true;
}

int main() {
    int s, k;
    cout << "Ab welcher Zahl sollen Primzahlen berechnet werden? " << endl;
    cin >> s;
    cout << "Bis zu welcher Zahl sollen Primzahlen berechnet werden? " << endl;
    cin >> k;
    cout << "Primzahlen von " << s << " bis " << k << ":" << endl;

    double start = clock();

    int count = 0;
    for (int i = s; i <= k; i++) {
        if (is_prime(i)) {
            count++;
        }
    }

    double end = clock();

    double percent = (double)count / (k - s + 1) * 100;

    cout << "Es wurden " << count << " Primzahlen gefunden. (" << (int)percent << "%)" << endl;
    double time_ms = (end - start) / CLOCKS_PER_SEC * 1000;
    cout << "Laufzeit: " << (int)time_ms << " ms" << endl;

    return 0;
}

File: Algorithmen/Primzahlen/py/Primzahlen_1bisK.py
import time
import math

def to_int(x):
    try:
        return int(x)
    except ValueError:
        for char in x:
            if char not in "0123456789.":
                x = x.replace(char, "")
            if x.count(".") > 1:
                x = x.replace(".", "")
            
            return int(x)
        
s = to_int(input("Ab welcher Zahl sollen Primzahlen berechnet werden? "))
k = to_int(input("Bis zu welcher Zahl sollen Primzahlen berechnet werden? "))

if s > k:
    s, k = k, s

def is_prime(n):
    if n < 2:
        return False
    for i in range(2, int(math.sqrt(n)+1)):
        if n % i == 0:
            return False
    return True

t1 = time.time()

primzahlen = [p for p in range(s, k+1) if is_prime(p)]

t2 = time.time()

print(f"ℙ ∈ [ {s}, {k} ]:")

print(", ".join(str(p) for p in primzahlen))

print(f"Es wurden {len(primzahlen)} Primzahlen gefunden. (" + str(int(len(primzahlen) / (k-s+1) * 100)) + " %)") 

print(f"Die Berechnung dauerte {int((t2 - t1)*1000)} ms.")